src/main/java/com/spikes2212/command/drivetrains/swerve/SwerveModule.java

```diff
@@ -0,0 +1,95 @@
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+
+public abstract class SwerveModule extends DashboardedSubsystem {
+
+    protected final SmartMotorController driveMotor;
+    protected final SmartMotorController turnMotor;
+
+    protected final boolean driveInverted;
+    protected final boolean turnInverted;
+    protected final double offset;
+
+    protected final PIDSettings drivePIDSettings;
+    protected final PIDSettings turnPIDSettings;
+    protected final FeedForwardSettings driveFeedForwardSettings;
+    protected final FeedForwardSettings turnFeedForwardSettings;
+
+    protected Rotation2d absoluteModuleAngle;
+
+    public SwerveModule(String namespaceName, SmartMotorController driveMotor, SmartMotorController turnMotor,
+                        boolean driveInverted, boolean turnInverted, double offset, PIDSettings drivePIDSettings,
+                        PIDSettings turnPIDSettings, FeedForwardSettings driveFeedForwardSettings,
+                        FeedForwardSettings turnFeedForwardSettings) {
+        super(namespaceName);
+        this.driveMotor = driveMotor;
+        this.turnMotor = turnMotor;
+        this.turnInverted = turnInverted;
+        this.driveInverted = driveInverted;
+        this.offset = offset;
+        this.drivePIDSettings = drivePIDSettings;
+        this.turnPIDSettings = turnPIDSettings;
+        this.driveFeedForwardSettings = driveFeedForwardSettings;
+        this.turnFeedForwardSettings = turnFeedForwardSettings;
+        driveMotor.setInverted(driveInverted);
+        turnMotor.setInverted(turnInverted);
+        configureTurnController();
+        configureDriveController();
+        configureAbsoluteEncoder();
+    }
+
+    public void setTargetState(SwerveModuleState targetState, double maxVelocity, boolean usePIDVelocity) {
+        targetState.optimize(Rotation2d.fromDegrees(turnMotor.getPosition()));
+        setTargetAngle(targetState.angle);
+        setTargetVelocity(targetState.speedMetersPerSecond, maxVelocity, usePIDVelocity);
+    }
+
+    public void setTargetAngle(Rotation2d targetAngle) {
+        turnMotor.pidSet(UnifiedControlMode.POSITION, targetAngle.getDegrees(), turnPIDSettings,
+                turnFeedForwardSettings, false);
+    }
+
+
+    public void setTargetVelocity(double targetVelocity, double maxVelocity, boolean usePIDVelocity) {
+        if (usePIDVelocity) {
+            driveMotor.pidSet(UnifiedControlMode.VELOCITY, targetVelocity, drivePIDSettings,
+                    driveFeedForwardSettings, false);
+        } else {
+            driveMotor.set(targetVelocity / maxVelocity);
+        }
+    }
+
+    protected abstract void configureDriveController();
+
+    protected abstract void configureTurnController();
+
+    protected abstract void configureAbsoluteEncoder();
+
+    protected abstract void updateAbsoluteModuleAngle();
+
+    public Rotation2d getAbsoluteModuleAngle() {
```

Comment:
This seems to have the same unnecesaity problem that `updateRobotAngle` has in `SwerveDrivetrain`. If I understand the code correctly, you can take a simpler approach and change this to be an abstract method that gets the current absoulte angle from the sensor